### ISSUE TITLE
Using LinkedHashMap's hashCode parameter seems to slow insertion

### ISSUE THREAD

**Issue by tarobins** [dart-lang/sdk#40340]
This tracker is for issues related to:
* Dart VM

If you aren't sure, file the issue here and we'll find the right home for it.
In your issue, please include:

* Dart SDK Version (`dart --version`)
Dart VM version: 2.8.0-dev.5.0 (Thu Jan 23 10:31:03 2020 +0100) on "macos_x64"

Using the `hashCode` parameter of the form `(i) => i` in LinkedHashMap's `hashCode` parameter on a Map of type <int, int> seems to slow the run time of 16m insertions by about 100ms.  I would expect since int's hashCode is just int, that returning the same from the `hashCode` param would run at the same speed.  

Running time data is here https://docs.google.com/spreadsheets/d/14AyxmYuwDdBrhiQY3E4lXUpHFnE1d972KzgAbMrsBEM/edit?usp=sharing

Code for generating this data is in [this branch](https://github.com/tarobins/map_timer/tree/hashCodeSlow).

**Comment by tatumizer** [comment]
This is just an added overhead of extra function call. (Try to call the function `(i)=>i` 15 million times in a loop - it takes approximately 100 milliseconds, which translates to 6 nanoseconds per call - quite a reasonable number).

**Comment by tarobins** [comment]
I don't think there should be an extra function call.  You're either calling `.hashCode` on the object being hashed or the function passed into the hash table.  In this case, both should be doing the same thing: returning the int as the hash code.  

It's possible when I'm using `.hashCode` on the object more is happening within the VM so a, possibly slow, call outside the VM is not needed.

I originally saw this issue on a pixel 2, where processing 9m items took an extra 500ms.

**Comment by tatumizer** [comment]
> You're either calling .hashCode on the object being hashed or the function passed into the hash table

int is not an object, it just pretends to be one. Most likely, it's special-cased by the implementation ("if the type of `key`  is `_Smi` then `hashCode=key` else ..." I can't find the source to say for sure, but the scenario looks quite plausible.

**Comment by lrhn** [comment]
No, `int` *is* an object, and the `HashMap` implementation does not special-case `int.hashCode`.

However, using the `equals` or `hashCode` arguments to the `Map` constructor means using a completely different implementation (`_CustomHashMap`) which is not as heavily optimized as `_HashMap`. For example, the custom hash mop starts each lookup with a check using the `_validKey` function, whether it was passed or not:
```dart
 V? operator [](Object? key) {
    if (!_validKey(key)) return null;
    ...
```
and that `_validKey` function defaults to doing a type check:
```dart
  _validKey = (validKey != null) ? validKey : new _TypeTest<K>().test
```
so you have one extra type check for each lookup.
That is necessary because the `operator[]` accepts any  object, but the `hashCode`  function provided here only works on `int`.

So, you have a very predictable overhead if you start using custom equals or hash-code computations.
